### PR TITLE
feat: add asn and asn_org field

### LINKED ISSUE #1: add asn field
A new table was created for asn. And the ability to import asn data was added.
Also, asn and as_org fields are available in query  response .

### PR DESCRIPTION
**Pull request by mohaphez**

### CODE REVIEW

File: src/Location.php [hunk 1]
```diff
@@ -134,14 +162,17 @@ public function setTimezone(string $timezone): void
...
 
     /**
-     * @return array{countryCode:string,subnet:string,country:string,region:string,city:string,latitude:float,longitude:float,zipcode:string,timezone:string}
+     * @return array{countryCode:string,subnet:string,country:string,countryCode:string,region:string,city:string,asn:int,asn_org:string,latitude:float,longitude:float,zipcode:string,timezone:string}
      */
     public function jsonSerialize(): array
     {
         $data = parent::jsonSerialize();
         $data['country'] = $this->country;
+        $data['country_eu'] = $this->countryCode;
```
Comment (yeganemehr) [inline src/Location.php]: this should indicate that the country is in [european union](https://ec.europa.eu/eurostat/statistics-explained/index.php?title=Glossary:Country_codes) or not.

Country|Code|Country|Code|Country|Code|Country|Code
-- | -- | -- | -- | -- | -- | -- | --
Belgium | BE | Greece | EL | Lithuania | LT | Portugal | PT
Bulgaria | BG | Spain | ES | Luxembourg | LU | Romania | RO
Czechia | CZ | France | FR | Hungary | HU | Slovenia | SI
Denmark | DK | Croatia | HR | Malta | MT | Slovakia | SK
Germany | DE | Italy | IT | Netherlands | NL | Finland | FI
Estonia | EE | Cyprus | CY | Austria | AT | Sweden | SE
Ireland | IE | Latvia | LV | Poland | PL

File: src/Location.php [hunk 2]
```diff
@@ -44,6 +48,8 @@ public static function fromSubnet($subnet): self
     public function __construct(
         string $countryCode,
         ICidr $subnet,
+        int $asn,
```
Comment (yeganemehr) [inline src/Location.php]: It's not nullable?

**Review by yeganemehr — CHANGES_REQUESTED** [review]

**Review by yeganemehr — APPROVED** [review]

File: database/migrations/2022_02_28_170001_create_subnets_v6_table.php [hunk 3]
```diff
@@ -22,6 +23,11 @@ public function up()
                 ->constrained('jeoip_ip2location_locations')
                 ->onUpdate('cascade')
                 ->onDelete('cascade');
+            $table->foreignIdFor(Asn::class, 'asn_id')
```
Comment (yeganemehr) [inline database/migrations/2022_02_28_170001_create_subnets_v6_table.php]: You're create asn foreign key before creating asn table!

Comment (yeganemehr) [inline database/migrations/2022_02_28_170001_create_subnets_v6_table.php]: Jeoip is deployed and running, we can't edit migrations that already ran on production.

File: src/Console/Commands/ImportData.php [hunk 4]
```diff
@@ -48,61 +79,63 @@ public function handle(): void
...
-            'zipcode',
-            'timezone',
-        ];
+        $columns = $this->option('asn') ?
+            $this->asnColumns :
+            $this->ip2LocationColumns;
+
         $count = count($data);
         if ($count != count($columns)) {
             throw new Exception('Csv data is not supported');
```
Comment (yeganemehr) [inline src/Console/Commands/ImportData.php:94]: I execute this command and after some time this exception thrown.

image

File: src/Console/Commands/ImportData.php [hunk 5]
```diff
@@ -121,7 +154,21 @@ protected function save(string $model, array $data)
...
+     * @param class-string<SubnetV4|SubnetV6> $model
+     * @param AsnCsvData                      $data
+     *
+     * @return SubnetV4|SubnetV6
+     */
+    protected function saveAsn(string $model, array $data)
+    {
+        $asn = Asn::updateOrCreate(['id' => $data['asn']], ['title' => $data['as']]);
+
+        return $model::where(['network_start' => $data['network_start'], 'network_end' => $data['network_end']])
```
Comment (yeganemehr) [inline src/Console/Commands/ImportData.php]: if you compare two files and db11.csv and asn.csv you can see network_start and network_end is not the same in both files.
image

_left is db11_ipv4.csv and other is asn_ipv4.csv_

**Review by yeganemehr — CHANGES_REQUESTED** [review]

File: database/migrations/2022_02_28_170001_create_subnets_v6_table.php (hunk 3 above)
Comment (mohaphez) [reply database/migrations/2022_02_28_170001_create_subnets_v6_table.php]: i removed that and create new migration file for add ASN column

File: src/Console/Commands/ImportData.php (hunk 5 above)
Comment (mohaphez) [reply src/Console/Commands/ImportData.php]: i have some changes in this method , please review that

File: database/migrations/2023_01_15_170001_create_asn_v4_table.php [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+return new class() extends Migration {
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::create('jeoip_ip2location_asn_v4', function (Blueprint $table) {
+            $table->id('id');
```
Comment (yeganemehr) [inline database/migrations/2023_01_15_170001_create_asn_v4_table.php]: every AS can have multiple network but based on implementation just last network kept.

**Review by yeganemehr — CHANGES_REQUESTED** [review]

**Review by yeganemehr — APPROVED** [review]

**Pull request merged**